### PR TITLE
RFC: Adding Measurement functionality to library

### PR DESCRIPTION
**Pull request by DGonzalezVillal**
Adding structures and functions to be able to calculate expected guest measurements.

### CODE REVIEW

File: src/measurement/sev_hashes.rs [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+            )));
+        }
+
+        let hashes_table = self.construct_table()?;
+        let mut page = Vec::with_capacity(4096);
+        page.resize(offset, 0);
+        page.extend_from_slice(&hashes_table[..]);
+        page.resize(4096, 0);
+        if page.len() != 4096 {
+            return Err(MeasurementError::SevHashError(SevHashError::InvalidSize(
```
Comment (larrydewey) [inline src/measurement/sev_hashes.rs:211]: Same nested error going on here.

File: src/measurement/vcpu_types.rs [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+//! Enum of different exisitng AMD EPYC VCPUs
+use std::{convert::TryFrom, fmt, str::FromStr};
+
+use crate::error::MeasurementError;
+
+/// All currently available QEMU VCPU types
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum CpuType {
+    /// EPYC
+    Epyc = 0,
```
Comment (larrydewey) [inline src/measurement/vcpu_types.rs:12]: I don't think you need to explicitly say 0, 1, 2, 3, etc, as we are not skipping anything. Those values  will be inferred.

File: src/measurement/vmsa.rs [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+        Self([T::default(); N])
+    }
+}
+
+/// SEV-ES VMSA page
+/// The names of the fields are taken from struct sev_es_work_area in the linux kernel:
+/// https://github.com/AMDESE/linux/blob/sev-snp-v12/arch/x86/include/asm/svm.h#L318
+/// (following the definitions in AMD APM Vol 2 Table B-4)
+#[repr(C)]
+#[derive(Debug, Clone, Copy, Serialize, Default)]
```
Comment (larrydewey) [inline src/measurement/vmsa.rs:122]: Do we need to derive deserialize?
```suggestion
#[derive(Debug, Clone, Copy, Serialize, Deserialize, Default)]
```

File: src/measurement/vmsa.rs [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+        area.rdx = rdx;
+        area.sev_features = sev_features;
+        area.xcr0 = 0x1;
+
+        area
+    }
+
+    /// Return a vector containing the save area pages
+    pub fn pages(&self, vcpus: usize) -> Result<Vec<Vec<u8>>, MeasurementError> {
+        let bsp_page = bincode::serialize(&self.bsp_save_area)
```
Comment (larrydewey) [inline src/measurement/vmsa.rs:350]: I believe we should be able to just use the `?` operator here instead of `map_err`

**Review by larrydewey — CHANGES_REQUESTED** [review]
This is some excellent work! There are some changes which I would like to discuss with you, but this is looking great!

File: src/measurement/gctx.rs [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+const ZEROS: [u8; LD_SIZE] = [0; LD_SIZE];
+
+/// Guest context field structure
+pub struct Gctx {
+    /// Launch Digest, 48 bytes long
+    ld: [u8; LD_SIZE],
+}
+
+/// Default init of GCTX, launch digest of all 0s
+impl Default for Gctx {
```
[the remark this answers is not in the document]
Comment (DGonzalezVillal) [reply src/measurement/gctx.rs:26]: Not supported for arrays with size bigger than 32.

File: src/measurement/gctx.rs [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    /// Initialize a new guest context using existing data
+    pub fn new(seed: &[u8]) -> Result<Self, MeasurementError> {
+        Ok(Self {
+            ld: seed.try_into()?,
+        })
+    }
+
+    /// Get the launch digest bytes
+    pub fn get_ld(self) -> [u8; LD_SIZE] {
+        self.ld
```
[the remark this answers is not in the document]
Comment (DGonzalezVillal) [reply src/measurement/gctx.rs:42]: Added reference, de-ref when used

File: src/measurement/ovmf.rs [hunk 7]
```diff
@@ -0,0 +1,320 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! Operations to handle ovmf data
+use crate::error::*;
+use bincode;
+use byteorder::{ByteOrder, LittleEndian};
+use serde::Deserialize;
```
[the remark this answers is not in the document]
Comment (DGonzalezVillal) [reply src/measurement/ovmf.rs:7]: OVMF will deserialize data, but not serialize any. We will serialize data in other modules.

File: src/measurement/ovmf.rs [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+    size: u32,
+    /// Version
+    version: u32,
+    /// Number of items
+    num_items: u32,
+}
+
+impl OvmfSevMetadataHeader {
+    /// Generate Header from bytes
+    fn from_bytes(value: &[u8], offset: usize) -> Result<Self, MeasurementError> {
```
[the remark this answers is not in the document]
Comment (DGonzalezVillal) [reply src/measurement/ovmf.rs:96]: Can't add macros without proc-macro. Created a trait instead.

File: src/measurement/ovmf.rs [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+    pub size: u32,
+    /// Section Type
+    pub section_type: SectionType,
+}
+
+impl OvmfSevMetadataSectionDesc {
+    /// Generate Description from bytes
+    fn from_bytes(value: &[u8], offset: usize) -> Result<Self, MeasurementError> {
+        let value = &value[offset..offset + std::mem::size_of::<OvmfSevMetadataSectionDesc>()];
+        bincode::deserialize(value).map_err(|e| MeasurementError::BincodeError(*e))
```
[the remark this answers is not in the document]
Comment (DGonzalezVillal) [reply src/measurement/ovmf.rs:76]: No

File: src/measurement/measurement_functions.rs [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+    match desc.section_type {
+        SectionType::SnpSecMemory => gctx.update_zero_pages(desc.gpa.into(), desc.size as usize)?,
+        SectionType::SnpSecrets => gctx.update_secrets_page(desc.gpa.into())?,
+        SectionType::CPUID => {
+            if vmm_type != VMMType::EC2 {
+                gctx.update_cpuid_page(desc.gpa.into())?
+            }
+        }
+        SectionType::SnpKernelHashes => {
+            snp_update_kernel_hashes(gctx, ovmf, sev_hashes, desc.gpa.into(), desc.size as usize)?
```
Comment (tylerfanelli) [inline src/measurement/measurement_functions.rs:65]: Shouldn't `sev_hashes` not be an `Option` at this point?

**Review by tylerfanelli — CHANGES_REQUESTED** [review]

File: src/lib.rs [hunk 11]
```diff
@@ -52,6 +52,8 @@ pub mod certs;
 pub mod firmware;
 #[cfg(target_os = "linux")]
 pub mod launch;
+#[cfg(all(any(feature = "sev", feature = "snp"), feature = "openssl"))]
```
[the remark this answers is not in the document]
Comment (larrydewey) [reply src/lib.rs:55]: There is measurement functionality for both SEV and SEV-SNP being introduced into the library. They are gated within the module.

File: src/measurement/gctx.rs [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+impl Gctx {
+    /// Initialize a new guest context using existing data
+    pub fn new(seed: &[u8]) -> Result<Self, MeasurementError> {
+        Ok(Self {
+            ld: seed.try_into()?,
+        })
+    }
+
+    /// Get the launch digest bytes
+    pub fn get_ld(self) -> [u8; LD_SIZE] {
```
[the remark this answers is not in the document]
Comment (DGonzalezVillal) [reply src/measurement/gctx.rs:41]: I think that is a good point, I can't think of a reason why you would want to call it until you're done updating the launch digest. Although the launch digest gets updated several times during the measurement calculation, I only call this function once the ld is completely updated. Should I still add a way of only being allowed to call this function once. @tylerfanelli. Thoughts @larrydewey?

File: src/measurement/gctx.rs [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+        }
+
+        self.ld = sha384(&page_info);
+
+        Ok(())
+    }
+
+    /// Update launch digest using normal memory pages
+    pub fn update_normal_pages(&mut self, start_gpa: u64, data: &[u8]) -> Result<(), GCTXError> {
+        if (data.len() % 4096) != 0 {
```
[the remark this answers is not in the document]
Comment (DGonzalezVillal) [reply src/measurement/gctx.rs:86]: @larrydewey were you referring to the size validation or the update pages function?

File: src/measurement/vmsa.rs (hunk 3 above)
Comment (DGonzalezVillal) [reply src/measurement/vmsa.rs:122]: At no point do I deserialize this struct, but I guess it won't hurt to add it.

File: src/measurement/vmsa.rs (hunk 4 above)
Comment (DGonzalezVillal) [reply src/measurement/vmsa.rs:350]: Same comment as bincode::deserialize. The error is boxed and so mapping it allows us to access the acutal error.

File: src/measurement/measurement_functions.rs (hunk 10 above)
Comment (DGonzalezVillal) [reply src/measurement/measurement_functions.rs:65]: What doe you mean?

Comment (tylerfanelli) [reply src/measurement/measurement_functions.rs:65]: Never mind, I now see that you update a zero page if `sev_hashes` is `None`.

File: src/measurement/measurement_functions.rs [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+        snp_update_section(desc, gctx, ovmf, sev_hashes, vmm_type)?
+    }
+
+    if vmm_type == VMMType::EC2 {
+        for desc in ovmf.metadata_items() {
+            if desc.section_type == SectionType::CPUID {
+                gctx.update_cpuid_page(desc.gpa.into())?
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (DGonzalezVillal) [reply src/measurement/measurement_functions.rs:89]: It seems that order of operation here matters. You need to first run through the metadata once and update everything, and then go through it again and only update the CPUID items if the instance is EC2. Grouping the logic together makes the test fail.

File: src/measurement/measurement_functions.rs [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+    let ovmf = OVMF::new(ovmf_file)?;
+    let mut gctx = Gctx::default();
+
+    gctx.update_normal_pages(ovmf.gpa(), ovmf.data())?;
+
+    Ok(gctx.get_ld())
+}
+
+/// Calulate an SEV-SNP launch digest
+#[allow(clippy::too_many_arguments)]
```
[the remark this answers is not in the document]
Comment (DGonzalezVillal) [reply src/measurement/measurement_functions.rs:111]: Sounds good!

File: src/lib.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (DGonzalezVillal) [reply src/lib.rs:55]: The SEV, SEV-ES and SNP measurements all differ. I put everything SNP related under snp and everything SEV/SEV-ES related under sev

[the remark this answers is not in the document]
Comment (tylerfanelli) [reply src/lib.rs:55]: Great, thanks for clearing that up.

File: src/measurement/gctx.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (larrydewey) [reply src/measurement/gctx.rs:41]: I think that is a reasonable thing to add. If we want to guarantee that it is only called when it is supposed to be, that would make sense.

File: src/measurement/measurement_functions.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (DGonzalezVillal) [reply src/measurement/measurement_functions.rs:89]: According to this [PR](https://github.com/virtee/sev-snp-measure/pull/13) in the python tooling:

"EC2 makes use a of a proprietary hypervisor that apparently loads the cpuid pages after all the other metadata pages"

So we do have to run it again when it's an EC2 machine.

File: Cargo.toml [hunk 16]
```diff
@@ -66,8 +66,9 @@ p384 = { version = "0.13.0", optional = true }
 rsa = { version = "0.9.2", optional = true }
 sha2 = { version = "0.10.8", optional = true }
 x509-cert = { version = "0.2.4", optional = true }
+byteorder = "1.4.3"
 
 [dev-dependencies]
 kvm-ioctls = ">=0.12"
 kvm-bindings = ">=0.6"
-serial_test = "2.0"
+serial_test = "2.0"
```
Comment (larrydewey) [inline Cargo.toml:74]: Not sure if this is an artifact of the merge conflict. Maybe double-check with main before adding this back in.

**Review by larrydewey — CHANGES_REQUESTED** [review]
Just a question about the Cargo.toml.

**Review by larrydewey — DISMISSED** [review]
LGTM

File: .gitignore [hunk 17]
```diff
@@ -1,2 +1,3 @@
 /target
 .vscode/
+Cargo.lock
```
Comment (tylerfanelli) [inline .gitignore:3]: `Cargo.lock` needs to be included in the git repository for the C API functionality with meson.

File: src/measurement/mod.rs [hunk 18]
```diff
@@ -0,0 +1,28 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! Everything one needs to calculate a launch measurement for a SEV encrypted confidential guest.
+//!
+//! This module contains the structures and functions needed to calculate a launch measurement.
+//! This includes, GCTX, SEV-HASHES, VMSA and OVMF pages.
+
+///
```
Comment (tylerfanelli) [inline src/measurement/mod.rs:8]:
```suggestion
```

File: src/measurement/ovmf.rs [hunk 19]
```diff
@@ -0,0 +1,334 @@
...
+        match value {
+            1 => Ok(SectionType::SnpSecMemory),
+            2 => Ok(SectionType::SnpSecrets),
+            3 => Ok(SectionType::CPUID),
+            0x10 => Ok(SectionType::SnpKernelHashes),
+            _ => Err(OVMFError::InvalidSectionType),
+        }
+    }
+}
+/// Creating strucutre from bytes
```
Comment (tylerfanelli) [inline src/measurement/ovmf.rs:59]:
```suggestion
/// Creating structure from bytes
```

File: src/measurement/ovmf.rs [hunk 20]
```diff
@@ -0,0 +1,334 @@
...
+pub trait TryFromBytes {
+    /// Error when attempting to deserialize from_bytes
```
Comment (tylerfanelli) [inline src/measurement/ovmf.rs:61]:
```suggestion
    /// Error when attempting to deserialize from bytes
```

File: src/measurement/ovmf.rs [hunk 21]
```diff
@@ -0,0 +1,334 @@
...
+    /// Size
+    size: u32,
+    /// Version
+    version: u32,
+    /// Number of items
+    num_items: u32,
+}
+
+impl TryFromBytes for OvmfSevMetadataHeader {
+    type Error = MeasurementError;
```
Comment (tylerfanelli) [inline src/measurement/ovmf.rs:104]:
```suggestion
    type Error = MeasurementError;
    
```

File: src/measurement/ovmf.rs [hunk 22]
```diff
@@ -0,0 +1,334 @@
...
+    /// Size
+    size: u16,
+    /// GUID
+    guid: [u8; 16],
+}
+
+impl TryFrom<&[u8]> for OvmfFooterTableEntry {
+    type Error = MeasurementError;
+
+    /// Grenerate footer from data
```
Comment (tylerfanelli) [inline src/measurement/ovmf.rs:140]:
```suggestion
    /// Generate footer from data
```

File: src/measurement/ovmf.rs [hunk 23]
```diff
@@ -0,0 +1,334 @@
...
+    /// OVMF data
+    data: Vec<u8>,
+    /// Table matching GUID to it's data
+    table: HashMap<Uuid, Vec<u8>>,
+    /// Metadata item description
+    metadata_items: Vec<OvmfSevMetadataSectionDesc>,
+}
+
+impl OVMF {
+    /// Generate new OVMF structure by parsing the foorter table and SEV metadata
```
Comment (tylerfanelli) [inline src/measurement/ovmf.rs:168]:
```suggestion
    /// Generate new OVMF structure by parsing the footer table and SEV metadata
```

File: src/measurement/ovmf.rs [hunk 24]
```diff
@@ -0,0 +1,334 @@
...
+const OVMF_TABLE_FOOTER_GUID: Uuid = uuid!("96b582de-1fb2-45f7-baea-a366c55a082d");
+const SEV_HASH_TABLE_RV_GUID: Uuid = uuid!("7255371f-3a3b-4b04-927b-1da6efa8d454");
+const SEV_ES_RESET_BLOCK_GUID: Uuid = uuid!("00f771de-1a7e-4fcb-890e-68c77e2fb44e");
+const OVMF_SEV_META_DATA_GUID: Uuid = uuid!("dc886566-984a-4798-a75e-5585a7bf67cc");
+
+/// OVMF Structure
+pub struct OVMF {
+    /// OVMF data
+    data: Vec<u8>,
+    /// Table matching GUID to it's data
```
Comment (tylerfanelli) [inline src/measurement/ovmf.rs:161]:
```suggestion
    /// Table matching GUID to its data
```

File: src/measurement/ovmf.rs [hunk 25]
```diff
@@ -0,0 +1,334 @@
...
+        if !(self.table.contains_key(&SEV_HASH_TABLE_RV_GUID)) {
+            return Err(OVMFError::EntryMissingInTable(
+                "SEV_HASH_TABLE_RV_GUID".to_string(),
+            ));
+        }
+        Ok(self
+            .table_item(&SEV_HASH_TABLE_RV_GUID)
+            .and_then(|entry| entry.get(..4))
+            .map(|bytes| LittleEndian::read_u32(bytes) as u64)
+            .unwrap())
```
Comment (tylerfanelli) [inline src/measurement/ovmf.rs:234]: Can we avoid the `unwrap()`?

File: src/measurement/ovmf.rs [hunk 26]
```diff
@@ -0,0 +1,334 @@
...
+    }
+
+    /// Get the SEV-ES EIP
+    pub fn sev_es_reset_eip(&self) -> Result<u32, OVMFError> {
+        if !(self.table.contains_key(&SEV_ES_RESET_BLOCK_GUID)) {
```
Comment (tylerfanelli) [inline src/measurement/ovmf.rs:239]:
```suggestion
        if !self.table.contains_key(&SEV_ES_RESET_BLOCK_GUID) {
```

File: src/measurement/ovmf.rs [hunk 27]
```diff
@@ -0,0 +1,334 @@
...
+            return Err(OVMFError::EntryMissingInTable(
+                "SEV_ES_RESET_BLOCK_GUID".to_string(),
+            ));
+        }
+
+        Ok(self
+            .table_item(&SEV_ES_RESET_BLOCK_GUID)
+            .and_then(|entry| entry.get(..4))
+            .map(LittleEndian::read_u32)
+            .unwrap())
```
Comment (tylerfanelli) [inline src/measurement/ovmf.rs:249]: Likewise, avoid the `unwrap()`.

File: src/measurement/sev_hashes.rs [hunk 28]
```diff
@@ -0,0 +1,214 @@
...
+/// SEV hash table entry
+#[repr(C)]
+#[derive(Debug, Clone, Copy, Serialize, Default)]
+struct SevHashTableEntry {
+    /// Entry GUID
+    guid: GuidLe,
+    /// Length
+    length: u16,
+    /// Hash
+    hash: Sha256Hash,
```
Comment (tylerfanelli) [inline src/measurement/sev_hashes.rs:61]:
```suggestion
    /// Entry GUID for ...
    guid: GuidLe,
    /// Length of ...
    length: u16,
    /// Hash representing ...
    hash: Sha256Hash,
```

A bit more detailed description of each struct member may be useful.

File: src/measurement/sev_hashes.rs [hunk 29]
```diff
@@ -0,0 +1,214 @@
...
+    guid: GuidLe,
+    /// Length
+    length: u16,
+    /// Cmd line append table entry
+    cmdline: SevHashTableEntry,
+    /// initrd table entry
+    initrd: SevHashTableEntry,
+    /// Kernel table entry
+    kernel: SevHashTableEntry,
+}
```
Comment (tylerfanelli) [inline src/measurement/sev_hashes.rs:88]:
```suggestion
/// Table of SEV hashes
#[repr(C)]
#[derive(Debug, Clone, Copy, Serialize, Default)]
struct SevHashTable {
    /// GUID of ...
    guid: GuidLe,
    /// Length of ...
    length: u16,
    /// Cmd line append table entry
    cmdline: SevHashTableEntry,
    /// initrd table entry
    initrd: SevHashTableEntry,
    /// Kernel table entry
    kernel: SevHashTableEntry,
}
```

Likewise, a bit more description of the struct members.

File: src/measurement/sev_hashes.rs [hunk 30]
```diff
@@ -0,0 +1,214 @@
...
+
+        Ok(SevHashes {
+            kernel_hash,
+            initrd_hash,
+            cmdline_hash,
+        })
+    }
+
+    ///Generate the SEV hashes area - this must be *identical* to the way QEMU
+    ///generates this info in order for the measurement to match.
```
Comment (tylerfanelli) [inline src/measurement/sev_hashes.rs:184]:
```suggestion
    /// Generate the SEV hashes area - this must be *identical* to the way QEMU
    /// generates this info in order for the measurement to match.
```

File: src/measurement/vcpu_types.rs [hunk 31]
```diff
@@ -0,0 +1,160 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! Enum of different exisitng AMD EPYC VCPUs
```
Comment (tylerfanelli) [inline src/measurement/vcpu_types.rs:3]:
```suggestion
//! Exisiting AMD EPYC vCPUs
```

File: src/measurement/vcpu_types.rs [hunk 32]
```diff
@@ -0,0 +1,160 @@
...
+use std::{convert::TryFrom, fmt, str::FromStr};
+
+use crate::error::MeasurementError;
+
+/// All currently available QEMU VCPU types
```
Comment (tylerfanelli) [inline src/measurement/vcpu_types.rs:8]:
```suggestion
/// All currently available QEMU vCPU types
```

File: src/measurement/vcpu_types.rs [hunk 33]
```diff
@@ -0,0 +1,160 @@
...
+    EpycIBPB,
+    /// EPYC V3
+    EpycV3,
+    /// EPYC V4
+    EpycV4,
+    /// EPYC ROME
+    EpycRome,
+    /// EPYC ROME V1
+    EpycRomeV1,
+    ///EPYC ROME V2
```
Comment (tylerfanelli) [inline src/measurement/vcpu_types.rs:27]:
```suggestion
    /// EPYC ROME V2
```

File: src/measurement/vmsa.rs [hunk 34]
```diff
@@ -0,0 +1,369 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! Operations to build and interact with an SEV-ES VMSA
+use crate::error::MeasurementError;
+// use crate::measurement::{measurement_functions::SevMode, vcpu_types::CpuType};
```
Comment (tylerfanelli) [inline src/measurement/vmsa.rs:5]:
```suggestion
```

File: src/measurement/vmsa.rs [hunk 35]
```diff
@@ -0,0 +1,369 @@
...
+        match s.to_lowercase().as_str() {
+            "sev" => Ok(SevMode::Sev),
+            "sev-es" | "seves" => Ok(SevMode::SevEs),
+            "sev-snp" | "sevsnp" => Ok(SevMode::SevSnp),
+            _ => Err(MeasurementError::InvalidSevModeError(s.to_string())),
+        }
+    }
+}
+
+/// VmmTypes
```
Comment (tylerfanelli) [inline src/measurement/vmsa.rs:35]:
```suggestion
/// Supported Virtual Machine Monitors
```

File: tests/measurement.rs [hunk 36]
```diff
@@ -0,0 +1,347 @@
+// SPDX-License-Identifier: Apache-2.0
+
+#![cfg(feature = "openssl")]
+
+#[cfg(feature = "snp")]
+mod snp_tests {
+    use sev::measurement::{snp::*, vmsa::VMMType};
+    // Test of we can generate a good OVMF hash
```
Comment (tylerfanelli) [inline tests/measurement.rs:8]:
```suggestion
    // Testing if we can generate a good OVMF hash
```

File: tests/measurement.rs [hunk 37]
```diff
@@ -0,0 +1,347 @@
...
+        let ld = snp_calc_launch_digest(arguments).unwrap();
+
+        let ld_hex = hex::encode(ld);
+
+        let exp_result = "db06fb267824b1ccb56edbe2a9c2ce88841bca5090dc6dac91d9cd30f3c2c0bf42daccb30d55d6625bfbf0dae5c50c6d";
+
+        assert_eq!(ld_hex.as_str(), exp_result);
+    }
+
+    // Test of we can a full LD from the OVMF hash
```
Comment (tylerfanelli) [inline tests/measurement.rs:33]:
```suggestion
    // Test if we can compute a full LD from the OVMF hash
```

**Review by tylerfanelli — CHANGES_REQUESTED** [review]

**Review by tylerfanelli — APPROVED** [review]

**Pull request merged**